internal/server/handlers_test.go

```diff
@@ -219,3 +219,87 @@ func TestPseudoDeltaSotWHandler(t *testing.T) {
 	require.WithinDuration(t, time.Now(), start.Add(wait), 10*time.Millisecond)
 	require.ElementsMatch(t, []*anypb.Any{testutils.MustMarshal(t, bazR).Resource}, lastRes.Resources)
 }
+
+func TestHandlerBatchingWithIRV(t *testing.T) {
+	var released atomic.Bool
+	ch := make(chan map[string]*ads.RawResource)
+	handler := newHandler(
+		testutils.Context(t),
+		NoopLimiter{},
+		NoopLimiter{},
+		new(customStatsHandler),
+		false,
+		func(resources map[string]*ads.RawResource) error {
+			ch <- maps.Clone(resources)
+			return nil
+		},
+	)
+	notify := func(name string, resource *ads.RawResource) {
+		handler.Notify(name, resource, ads.SubscriptionMetadata{})
+	}
+
+	expectedEntries := make(map[string]*ads.RawResource)
+
+	t.Run("partial update, foo not updated and bar updated", func(t *testing.T) {
+		req := newDeltaReq([]string{"foo", "bar"}, map[string]string{"foo": "0", "bar": "0"})
+		handler.StartNotificationBatch(req.InitialResourceVersions)
+		fooResource := newRawResource("foo", "0")
+		barResource := newRawResource("bar", "1")
+		expectedEntries["bar"] = barResource
+
+		notify("foo", fooResource)
+		notify("bar", barResource)
+
+		released.Store(true)
+		handler.EndNotificationBatch()
+		require.Equal(t, expectedEntries, <-ch)
+		clear(expectedEntries)
+	})
+
+	t.Run("partial update, foo deleted and bar updated", func(t *testing.T) {
+		req := newDeltaReq([]string{"foo", "bar"}, map[string]string{"foo": "0", "bar": "0"})
+		handler.StartNotificationBatch(req.InitialResourceVersions)
+		barResource := newRawResource("bar", "0")
+		notify("foo", nil)
+		notify("bar", barResource)
+		expectedEntries["foo"] = nil
+		released.Store(true)
+		handler.EndNotificationBatch()
+		require.Equal(t, expectedEntries, <-ch)
+
+		clear(expectedEntries)
+	})
+
+	t.Run("partial update, foo deleted, bar updated and foo re-created", func(t *testing.T) {
+		req := newDeltaReq([]string{"foo", "bar"}, map[string]string{"foo": "0", "bar": "0"})
+		handler.StartNotificationBatch(req.InitialResourceVersions)
+
+		newFooResource := newRawResource("foo", "1")
+		barResource := newRawResource("bar", "1")
+		notify("foo", nil)
```

Comment:
This won't really happen IRL, unless `foo` gets set while the subscriptions are happening. It's not really representative of real scenarios. You can just remove this line and the test doesn't really change.